### DISCUSSION TITLE
Langchain tools / CrewAI Intergration

### DISCUSSION THREAD

**Discussion by Kingbadger3d** [open-webui/open-webui#3805] [Ideas]
Has anyone been looking at adding Langchain tool support, e.g web scraper, internet search etc etc for agent use. Also crewai intergration which uses langchain anyway would be a big advancement with the tool. Agents rule this space, we need support asap. Cheers KB

**Comment by tjbck** [comment]
Would love to see more contribution in this space with Tools & Functions, it's 100% doable!

**Comment by Kingbadger3d** [comment]
Ok, lets get this started. Ive been playing with a system just released white paper and code example, hacked it to use ollama amd grok i should be able to port that over fairly easy from what ive been doing.

Here's the repo: https://github.com/togethercomputer/moa

Read the paper this is beating GPT4omega by 10% in benchmarks. Ive alwats said multi architecture multi agent systems should be supperior to 99% of frontier solutions.

heres a provided single Python file version of the system, this can be changed to be used in the function area of latest openwebui with tweaks, this code is 100% there's on the repo:

# Mixture-of-Agents in 50 lines of code
import asyncio
import os
from together import AsyncTogether, Together

client = Together(api_key=os.environ.get("TOGETHER_API_KEY"))
async_client = AsyncTogether(api_key=os.environ.get("TOGETHER_API_KEY"))

user_prompt = "What are some fun things to do in SF?"
reference_models = [
    "Qwen/Qwen2-72B-Instruct",
    "Qwen/Qwen1.5-72B-Chat",
    "mistralai/Mixtral-8x22B-Instruct-v0.1",
    "databricks/dbrx-instruct",
]
aggregator_model = "mistralai/Mixtral-8x22B-Instruct-v0.1"
aggreagator_system_prompt = """You have been provided with a set of responses from various open-source models to the latest user query. Your task is to synthesize these responses into a single, high-quality response. It is crucial to critically evaluate the information provided in these responses, recognizing that some of it may be biased or incorrect. Your response should not simply replicate the given answers but should offer a refined, accurate, and comprehensive reply to the instruction. Ensure your response is well-structured, coherent, and adheres to the highest standards of accuracy and reliability.

Responses from models:"""


async def run_llm(model):
    """Run a single LLM call with a reference model."""
    response = await async_client.chat.completions.create(
        model=model,
        messages=[{"role": "user", "content": user_prompt}],
        temperature=0.7,
        max_tokens=512,
    )
    print(model)
    return response.choices[0].message.content


async def main():
    results = await asyncio.gather(*[run_llm(model) for model in reference_models])

    finalStream = client.chat.completions.create(
        model=aggregator_model,
        messages=[
            {"role": "system", "content": aggreagator_system_prompt},
            {"role": "user", "content": ",".join(str(element) for element in results)},
        ],
        stream=True,
    )

    for chunk in finalStream:
        print(chunk.choices[0].delta.content or "", end="", flush=True)


asyncio.run(main())

This system mixed with Langchain and an agent system like crewai with be baller.

**Comment by rodeldoria** [comment]
I tried to do crewai with my Gpt4 api and it would burn through my credits. It would make sense to run it all with Ollama then use gpt4o to summarize it. I am going to be using web scraping and other tools to run my site.
